### ISSUE TITLE
Support stateful session affinity for weighted clusters to support use-cases like canary deployment

### ISSUE THREAD

**Issue by sanjaypujare** [envoyproxy/envoy#24741]
*Title*: Support stateful session affinity for weighted clusters to support use-cases like canary deployment

*Description*:
Currently cookie based stateful session affinity is implemented (see https://github.com/envoyproxy/envoy/pull/18207) but it only works within a cluster. When multiple clusters are used for a service and the user would like to maintain session affinity across clusters then it doesn't work. 

e.g. client sends a request R1 and Envoy routes it to cluster C1 host C1H1. Envoy inserts a cookie in the response to maintain session affinity to H1.  Client copies the cookie value in the next request R2 and because weighted-clusters (canary deployment) is in use, Envoy picks cluster C2 for R2 but the cookie is meaningless for C2 so Envoy picks a new host C2H2 and session affinity is broken. What we want is that Envoy should pick cluster C1 for R2 (because the cookie also has the cluster name encoded in it) and within C1 it should route R2 to C1H1.

[optional *Relevant Links*:]
https://github.com/istio/api/pull/2469#issuecomment-1245067340
https://github.com/istio/api/pull/2469#issuecomment-1246231933
https://github.com/istio/api/pull/2469#issuecomment-1247587910
https://github.com/istio/api/pull/2469#issuecomment-1248422947

CC @wbpcode

**Comment by wbpcode** [comment]
sound reasonable requirement.

**Comment by sanjaypujare** [comment]
> sound reasonable requirement.

Any plans to address it? I am asking based on your comments in https://github.com/istio/api/pull/2469

We have users who are requesting this feature in Envoy.

**Comment by wbpcode** [comment]
Sorry, I have no bandwidth for this temporarily. I can add this to my TODO list.
Or are you willing or have time to do some contributions?

**Comment by sanjaypujare** [comment]
> Sorry, I have no bandwidth for this temporarily. I can add this to my TODO list. Or are you willing or have time to do some contributions?

No, I cannot do this either. Let's see if we can find some volunteer

**Comment by wbpcode** [comment]
ooh, thanks very much for you taking over this @adisuissa. I added two comments to the way to mutate the cluster, and the format.

**Comment by Tratcher** [comment]
I also need this in my service. @adisuissa, any progress, or should I try to do it?

**Comment by markdroth** [comment]
Note that gRPC has already implemented this feature; see [gRFC A60](https://github.com/grpc/proposal/blob/master/A60-xds-stateful-session-affinity-weighted-clusters.md) for the design.

As discussed in #30678, we should keep the Envoy impl compatible with gRPC.
